examples/samv/blink/main.cpp

```diff
@@ -0,0 +1,58 @@
+/*
+ * Copyright (c) 2021, Tomasz Wasilczyk
+ *
+ * This file is part of the modm project.
+ *
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+
+#include <modm/architecture/interface/clock.hpp>
+#include <modm/platform.hpp>
+
+using namespace std::chrono_literals;
+using namespace modm::platform;
+
+struct SystemClock
+{
+	// Chosen to achieve 300MHz system clock
+	static constexpr uint32_t PllAMult = 25;
+	static constexpr uint32_t Frequency = 300'000'000;  // TODO: PllAMult * 12'000'000 (OSC12M);
+
+	static bool inline
+	enable()
+	{
+		// TODO: enable OSC12M?
+		// TODO: enable MAINCK?
+		ClockGen::setFlashLatency<Frequency>();  // TODO: what is this for?
+		ClockGen::updateCoreFrequency<Frequency>();  // TODO: what is this for?
```

Comment:
Oh, and also, I meant to point out: You have set frequency to 300M here, but then you set the master clock to MAIN_CLK, which I believe will be 12M. So your delays will be  300/12=25x longer than you expect.

Reply:
How do I tune this? Just by measurements, or from some source? I tried changing these numbers to extreme values (1 or 24) and it doesn't seem to affect 500ms delays in my sample code.

Indeed MAIN_CLK is 12MHz and delays should be 25x, but apparently they are just off by ~2x (I still need to measure this). I tried changing this to PLLA, but it doesn't boot then. Seems like I have some heavy debugging to be done here.

Reply:
> I tried changing these numbers to extreme values (1 or 24) and it doesn't seem to affect 500ms delays in my sample code.

Because the Cortex-M7 implementation uses a hardware counter. This is relevant for `modm::delay_ns`, which is done in software. The correct value is probably 3 or 4. You can tune this by running the `generic/delay` example until things work.

I've also [written about this implementation here](https://blog.salkinium.com/modm-delay/).

Reply:
I measured the delay to calibrate it with a scope and a togglng GPIO.

So, actually `delay_us` and `delay_ms` uses the DWT cycle counter -- on platforms with DWT, which I believe includes yours (I think all by M0) -- to implement delays, rather than busy looping, so in that case the parameters there won't matter. However, `delay_ns` uses a decrement loop, and that's where the calibration factors get used. It looks like if you call `modm::delay(10000ns)`, `delay_ns` gets used. With `modm::delay(500ms)` it uses the delay_ms which probably uses DWT. Of course, you can also just call `delay_ns(x)`

Reply:
Finally, I've been able to configure master clock off of PLLA. The problem was indeed MCK/HCLK divider - without it, program hanged after setting up master clock. Now, small delays are *pretty much working +/- 10% (what's okay given that I'm using internal uncalibrated oscillator). Thanks for all the input!

* - with one problem, described in a separate comment

Reply:
I configured wait states with some margin so that they fit SAMG as well. I still need to account for MCLK!=HCLK.

Tweaking delay_us delays is the next thing to do, but I'm afraid I won't do it very precisely without external crystal.

I sorted out most of the TODOs, there's 4 left. The change is almost ready for final review.

Reply:
I just calibrated the first value for delay_ns and it turned out to be **1**. I can't find out how to calibrate the second one - the minimal delay I can get is around ~200ns.

Reply:
I think with the most recent revision I have addressed all the comments and TODOs from this thread. Thanks!